windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1348 @@
+    ['{7442A323-C1E3-404B-BEEA-5B24A52BBB0E}']
+    function Params: TUpdateStateMachineParams;
+  end;
+
+  TOnlineUpdateSharedData = class(TInterfacedObject, IOnlineUpdateSharedData)
+  private
+    FParams: TUpdateStateMachineParams;
+  public
+    constructor Create(AParams: TUpdateStateMachineParams);
+    function Params: TUpdateStateMachineParams;
+  end;
+  // Private Utility functions
+  function ConfigCheckContinue: Boolean;
+implementation
+
+uses
+  Winapi.Shlobj,
+  System.WideStrUtils,
+  Vcl.Dialogs,
+  Winapi.ShellApi,
+  Winapi.Windows,
+  Winapi.WinINet,
+
+  GlobalProxySettings,
+  KLog,
+  keymanapi_TLB,
+  KeymanVersion,
+  kmint,
+  ErrorControlledRegistry,
+  RegistryKeys,
+  Upload_Settings,
+  utildir,
+  utilexecute,
+  OnlineUpdateCheckMessages,    // todo create own messages
+  UfrmOnlineUpdateIcon,
+  UfrmOnlineUpdateNewVersion,
+  utilsystem,
+  utiluac,
+  versioninfo,
+  Keyman.System.RemoteUpdateCheck,
+  Keyman.System.DownloadUpdate;
+
+const
+  SPackageUpgradeFilename = 'upgrade_packages.inf';
+  kmShellContinue = 0;
+  kmShellExit = 1;
+
+{ TUpdateStateMachine }
+
+constructor TUpdateStateMachine.Create(AForce : Boolean);
+begin
+  inherited Create;
+  FShowErrors := True;
+  FParams.Result := oucUnknown;
+
+  FForce := AForce;
+  FAutomaticUpdate := GetAutomaticUpdates;
+  FIdle := IdleState.Create(Self);
+  FUpdateAvailable := UpdateAvailableState.Create(Self);
+  FDownloading := DownloadingState.Create(Self);
+  FWaitingRestart := WaitingRestartState.Create(Self);
+  FInstalling := InstallingState.Create(Self);
+  FRetry := RetryState.Create(Self);
+  FPostInstall := PostInstallState.Create(Self);
+  // Check the Registry setting.
+  SetStateOnly(ConvertEnumState(CheckRegistryState));
+end;
+
+destructor TUpdateStateMachine.Destroy;
+begin
+  if (FErrorMessage <> '') and FShowErrors then
+    KL.Log(FErrorMessage); // TODO: #10210 Log to Sentry
+
+  if FParams.Result = oucShutDown then
+    ShutDown;
+
+  FIdle.Free;
+  FUpdateAvailable.Free;
+  FDownloading.Free;
+  FWaitingRestart.Free;
+  FInstalling.Free;
+  FRetry.Free;
+  FPostInstall.Free;
+
+  // TODO: #10210 remove debugging comments
+  //KL.Log('TUpdateStateMachine.Destroy: FErrorMessage = '+FErrorMessage);
+  //KL.Log('TUpdateStateMachine.Destroy: FParams.Result = '+IntToStr(Ord(FParams.Result)));
+
+  inherited Destroy;
+end;
+
+
+procedure TUpdateStateMachine.SavePackageUpgradesToDownloadTempPath;
+var
+  i: Integer;
+  StringList : TStringList;
+begin
+  StringList := TStringList.Create;
+  try
+    for i := 0 to High(FParams.Packages) do
+      if FParams.Packages[i].NewID <> '' then
+        StringList.Add(FParams.Packages[i].NewID+'='+FParams.Packages[i].ID);
+    if StringList.Count > 0 then
+      StringList.SaveToFile(DownloadTempPath + SPackageUpgradeFileName);
+  finally
+    StringList.Free;
+  end;
+end;
+
+procedure TUpdateStateMachine.ShutDown;
+begin
+  if Assigned(Application) then
+    Application.Terminate;
```

Comment:
This puts a dependency on VCL in a System unit, which is not ideal. We normally want System units to have no VCL deps. It would be better for this to have a callback so its own can handle application shutdown cleanly in its own way (whether that be Application.Terminate or something else).

Reply:
I have remove the VCL but still have to forms that pop for start install on a restart, and start Install Now.